sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpMessageBatch.cs

```diff
@@ -154,14 +154,14 @@ public override void Clear()
         ///
         /// <returns>The set of messages as an enumerable of the requested type.</returns>
         ///
-        public override IEnumerable<T> AsEnumerable<T>()
+        public override IReadOnlyCollection<T> AsReadOnly<T>()
```

Comment:
We only used `IEnumerable` because the legacy client used it and we didn't reevaluate after mimicking the pattern.  Should we consider just returning the source type (List<T>) here and avoid any cast?   Its internal use and we have no plans to expose it wider and the callers are trusted.   On the other hand, this form does protect us from accidental mutation.    Thoughts?

Reply:
At first I had to leave it with `IReadOnlyCollection` because there was an overload that passed in an array with a single message. Now that I have introduced an overload that accepts just one message and allows bypassing that additional array allocation, it would be possible to just expose the `List<T>` directly as you mentioned. 

From a performance perspective, it would be slightly better because we can make sure the right enumerator is used. Not sure how powerful the latest devirtualization improvements are and whether it matters that much. I'd personally also buy into the trusted caller perspective and go without the abstraction. I can see also the benefit of the abstraction since the data returned there is consumed by multiple internal method calls.

Reply:
Ah, I missed this discussion. Let me take a look and get back to you on this.

Reply:
I think we can go ahead with the change to `IList<T>`. 

One thing I noticed (and it looks like Jesse is already handling in his EH PR) is that if we do this, we would need to still have the SendBatchInternalAsync method in AmqpSender take an IEnumerable to handle the [public API](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/servicebus/Azure.Messaging.ServiceBus/src/Sender/ServiceBusSender.cs#L193) that takes an IEnumerable events which we then convert to a readonlylist. It looks like in Jesse's PR he is adding a new overload to [AmqpMessageConverter](https://github.com/Azure/azure-sdk-for-net/pull/26983/files#diff-da252bede9fe91fffbfd42d0f1f8df3a051fa54640d936a71602535fce9b10bcR82) so that we can still benefit from the perf enhancements for the MessageBatch case.

Reply:
Agreed; I had forgotten that we were eagerly materializing the enumerable at the call site.